### PR TITLE
Added CSP Middleware draft

### PR DESCRIPTION
**Pull request by jrson83**
## Reference Issues/PRs

[https://github.com/lumeland/lume/issues/183](https://github.com/lumeland/lume/issues/183)

## What does this implement/fix?

New Middleware for Security Headers in production.

Todo:
- [x] Strict-Transport-Security
- [x] Content-Security-Policy
  - [x] Report-Only
  - [x] nonce
  - [x] ~sha-256~
- [x] Referrer-Policy
- [x] Permissions-Policy (will use `string` for now)
- [x] Expect-CT (will be deprecated soon, so disabled by default)
- [x] X-Frame-Options
- [x] X-Content-Type-Options
- [x] X-XSS-Protection
- [x] X-Permitted-Cross-Domain-Policies
- [x] X-Powered-By
- [ ] Create default options (best practices)
- [ ] Add functionality to receive csp-reports (server side)

### Related tools
- [securityheaders.com](https://securityheaders.com/)
- [validbot.com](https://www.validbot.com/)
- [CSP Evaluator](https://csp-evaluator.withgoogle.com/)

### Testing
When testing make sure to not use the dev server, since the headers will be set but won't work.

Instead use prod server e.g.:

```ts
// deno run --allow-net --allow-read serve.ts
import Server from "lume/core/server.ts";
import csp from "./csp.ts"

const server = new Server({
  port: 8000,
  root: `${Deno.cwd()}/_site`,
});

server.use(csp());

server.start();

console.log("Listening on http://localhost:8000");
```

### CODE REVIEW

File: csp/mod.ts [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+export const defaults: Options = {
+  "Strict-Transport-Security": {
+    "max-age": DEFAULT_MAX_AGE,
+    "includeSubDomains": true,
+    "preload": true,
+  },
+  "Content-Security-Policy": {
+    "default-src": ["'none'"],
+    "script-src": ["cdn.jsdelivr.net", "'self'"],
+    "style-src": ["cdn.jsdelivr.net", "'self'"],
```
Comment (oscarotero) [inline csp/mod.ts:114]: I'm not sure about including some domains by default. I think this should be configured by the user, and leave only "self".

File: csp/mod.ts [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+  "Expect-CT": {
+    "max-age": DEFAULT_MAX_AGE,
+    "enforce": true,
+    "report-uri": "https://localhost:8000/report/",
+  },
+  "X-Frame-Options": true,
+  "X-Content-Type-Options": true,
+  "X-XSS-Protection": true,
+  "X-Permitted-Cross-Domain-Policies": true,
+  "X-Powered-By": "Fake Server",
```
Comment (oscarotero) [inline csp/mod.ts:129]: The `X-Powered-By` should be disabled by default.

File: csp/mod.ts [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+function getExpectCt(options: Readonly<ExpectCtOptions>): string {
+  const headerValue: string[] = [
+    `max-age=${validateMaxAge(options["max-age"])}`,
+  ];
+
+  if (options.enforce) {
+    headerValue.push("enforce");
+  }
+
+  if (options["enforce"]) {
```
Comment (oscarotero) [inline csp/mod.ts:283]: I guess you mean "report-uri" here :)

**Review by oscarotero — CHANGES_REQUESTED** [review]

File: csp/mod.ts (hunk 3 above)
Comment (jrson83) [reply csp/mod.ts:283]: Right. Fixed.

File: csp/mod.ts [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+    "style-src": ["cdn.jsdelivr.net", "'self'"],
+    "font-src": ["'self'"],
+    "img-src": ["shield.deno.dev", "'self'"],
+  },
+  "Referrer-Policy": ["no-referrer", "strict-origin-when-cross-origin"],
+  "Permissions-Policy": "camera=(), microphone=(), geolocation=()",
+  "Expect-CT": {
+    "max-age": DEFAULT_MAX_AGE,
+    "enforce": true,
+    "report-uri": "https://localhost:8000/report/",
```
[the remark this answers is not in the document]
Comment (jrson83) [reply csp/mod.ts:123]: See comment `All current default values are just for testing purpose with lume.land while dev`

File: csp/mod.ts (hunk 1 above)
Comment (jrson83) [reply csp/mod.ts:114]: See comment `All current default values are just for testing purpose with lume.land while dev`

File: csp/mod.ts (hunk 2 above)
Comment (jrson83) [reply csp/mod.ts:130]: Yes, I just added a test value, for testing purpose, to see if it works if a string is set.

File: csp/mod.ts [hunk 5]
```diff
@@ -0,0 +1,288 @@
+import type { Middleware, RequestHandler } from "lume/core.ts";
+import { merge } from "lume/core/utils.ts";
+
+const DEFAULT_MAX_AGE = 365 * 86400;
+
+type StrictTransportSecurityOptions = {
```
[the remark this answers is not in the document]
Comment (jrson83) [reply csp/mod.ts:6]: Agree. Will do.

File: csp/mod.ts [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+  "max-age": number;
```
[the remark this answers is not in the document]
Comment (jrson83) [reply csp/mod.ts:7]: Agree. Will do.

File: csp/mod.ts [hunk 7]
```diff
@@ -1,46 +1,147 @@
...
+   * Deprecated:
+   * Prevents loading any assets over HTTP when the page uses HTTPS
+   */
+  "block-all-mixed-content"?: string[];
+
+  /**
+   * Deprecated:
+   * Used to specify information in the Referer header for links away from a page
+   */
+  "referrer"?: [
```
Comment (oscarotero) [inline csp/mod.ts:109]: I think this type is wrong. "referrer" accept only one of these values, right? It should be a enum:
```
"referrer"?: "no-referrer" | "no-referrer-when-downgrade" | "origin" | "origin-when-cross-origin" | "unsafe-url";
```

**Review by oscarotero — CHANGES_REQUESTED** [review]

File: csp/mod.ts (hunk 7 above)
Comment (jrson83) [reply csp/mod.ts:109]: right.

**Pull request merged**